Add padding: 1px to <td> and <th> for make spec-compliant

AlexsanderDamaceno opened the pull request:
#### 83eaa21149af78ac78da917d3fc421dda3282f8c
<pre>
Add padding:1px to &lt;td&gt; and &lt;th&gt; for make spec-compliant
<a href="https://bugs.webkit.org/show_bug.cgi?id=255844">https://bugs.webkit.org/show_bug.cgi?id=255844</a>

Reviewed by NOBODY (OOPS!).

Add padding:1px to &lt;td&gt; and &lt;th&gt; to html.css.
Fix wpt test that was failing because of the missing padding.

* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/html-to-css-mapping-1-expected.txt:
* Source/WebCore/css/html.css:
(td, th):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/83eaa21149af78ac78da917d3fc421dda3282f8c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78678 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57722 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32060 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83338 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29940 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80811 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66874 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6003 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61618 "Found 223 new test failures: css1/box_properties/float_elements_in_series.html css2.1/t0905-c5525-fltmult-00-d-g.html css2.1/tables/table-anonymous-objects-091.xht css2.1/tables/table-anonymous-objects-092.xht css2.1/tables/table-anonymous-objects-099.xht css2.1/tables/table-anonymous-objects-100.xht css2.1/tables/table-anonymous-objects-103.xht css2.1/tables/table-anonymous-objects-104.xht css2.1/tables/table-anonymous-objects-105.xht css2.1/tables/table-anonymous-objects-106.xht ... (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19541 "Found 3 new test failures: fast/css-grid-layout/grid-item-overflow-paint.html fast/dom/HTMLTableElement/cellpadding-attribute.html fast/forms/005.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81745 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51649 "Found 60 new test failures: compositing/contents-format/deep-color-backing-store.html css1/box_properties/float_elements_in_series.html css2.1/t0905-c5525-fltmult-00-d-g.html css2.1/tables/table-anonymous-objects-091.xht css2.1/tables/table-anonymous-objects-092.xht css2.1/tables/table-anonymous-objects-099.xht css2.1/tables/table-anonymous-objects-100.xht css2.1/tables/table-anonymous-objects-103.xht css2.1/tables/table-anonymous-objects-104.xht css2.1/tables/table-anonymous-objects-105.xht ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/70504 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41927 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48995 "Found 15 new test failures: imported/w3c/web-platform-tests/css/css-break/table/table-rowspan-001.html imported/w3c/web-platform-tests/css/css-tables/colspan-001.html imported/w3c/web-platform-tests/css/css-tables/colspan-002.html imported/w3c/web-platform-tests/css/css-tables/colspan-003.html imported/w3c/web-platform-tests/css/css-tables/colspan-004.html imported/w3c/web-platform-tests/css/css-tables/height-distribution/computing-row-measure-0.html imported/w3c/web-platform-tests/css/css-tables/height-distribution/computing-row-measure-1.html imported/w3c/web-platform-tests/css/css-tables/percent-width-ignored-001.tentative.html imported/w3c/web-platform-tests/css/css-tables/percent-width-ignored-002.tentative.html imported/w3c/web-platform-tests/css/css-tables/percent-width-ignored-003.tentative.html ... (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25589 "Found 60 new test failures: css1/box_properties/float_elements_in_series.html css2.1/t0905-c5525-fltmult-00-d-g.html css2.1/tables/table-anonymous-objects-091.xht css2.1/tables/table-anonymous-objects-092.xht css2.1/tables/table-anonymous-objects-099.xht css2.1/tables/table-anonymous-objects-100.xht css2.1/tables/table-anonymous-objects-103.xht css2.1/tables/table-anonymous-objects-104.xht css2.1/tables/table-anonymous-objects-105.xht css2.1/tables/table-anonymous-objects-106.xht ... (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28280 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70100 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25967 "Found 60 new test failures: css1/box_properties/float_elements_in_series.html css2.1/t0905-c5525-fltmult-00-d-g.html css2.1/tables/table-anonymous-objects-091.xht css2.1/tables/table-anonymous-objects-092.xht css2.1/tables/table-anonymous-objects-099.xht css2.1/tables/table-anonymous-objects-100.xht css2.1/tables/table-anonymous-objects-103.xht css2.1/tables/table-anonymous-objects-104.xht css2.1/tables/table-anonymous-objects-105.xht css2.1/tables/table-anonymous-objects-106.xht ... (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84707 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6043 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4162 "Found 60 new test failures: css1/box_properties/float_elements_in_series.html css2.1/t0905-c5525-fltmult-00-d-g.html css2.1/tables/table-anonymous-objects-091.xht css2.1/tables/table-anonymous-objects-092.xht css2.1/tables/table-anonymous-objects-099.xht css2.1/tables/table-anonymous-objects-100.xht css2.1/tables/table-anonymous-objects-103.xht css2.1/tables/table-anonymous-objects-104.xht css2.1/tables/table-anonymous-objects-105.xht css2.1/tables/table-anonymous-objects-106.xht ... (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69844 "Failure limit exceed. At least found 214 new test failures: css1/box_properties/float_elements_in_series.html css2.1/t0905-c5525-fltmult-00-d-g.html css2.1/tables/table-anonymous-objects-091.xht css2.1/tables/table-anonymous-objects-092.xht css2.1/tables/table-anonymous-objects-099.xht css2.1/tables/table-anonymous-objects-100.xht css2.1/tables/table-anonymous-objects-103.xht css2.1/tables/table-anonymous-objects-104.xht css2.1/tables/table-anonymous-objects-105.xht css2.1/tables/table-anonymous-objects-106.xht ... (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6205 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67624 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69099 "Found 3 new API test failures: /TestWebKit:WebKit.HitTestResultNodeHandle, /TestWebKit:WebKit.FrameMIMETypePNG, /TestWebKit:WebKit.WillSendSubmitEvent (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13136 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11669 "Found 60 new test failures: compositing/overflow/overflow-scroll-to-visible.html css1/box_properties/float_elements_in_series.html css2.1/t0905-c5525-fltmult-00-d-g.html css2.1/tables/table-anonymous-objects-091.xht css2.1/tables/table-anonymous-objects-092.xht css2.1/tables/table-anonymous-objects-099.xht css2.1/tables/table-anonymous-objects-100.xht css2.1/tables/table-anonymous-objects-103.xht css2.1/tables/table-anonymous-objects-104.xht css2.1/tables/table-anonymous-objects-105.xht ... (failure)") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5988 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/11910 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5974 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9410 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7762 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/559616676397fab2e59d22320bab095e0a04f4f2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37269

AlexsanderDamaceno committed:
Add padding:1px to <td> and <th> for make spec-compliant
https://bugs.webkit.org/show_bug.cgi?id=255844

Reviewed by NOBODY (OOPS!).

Add padding:1px to <td> and <th> to html.css.
Fix wpt test that was failing because of the missing padding.

* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/html-to-css-mapping-1-expected.txt:
* Source/WebCore/css/html.css:
(td, th):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5eefb0072a675409a807c7f45d637074de97e9d4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37269

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/83eaa21149af78ac78da917d3fc421dda3282f8c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37269

nt1m commented:
A lot of tests are now failing.

You can add the `no-failure-limits` tag to the PR and use `update-test-expectations github-pr`. It will do most of the job (but not all of it).

karlcow commented:
@AlexsanderDamaceno are you planning to make progress on this?

karlcow commented:
Simple test to help too. 
https://codepen.io/webcompat/pen/azoXwxm

karlcow commented on Source/WebCore/css/html.css:
```diff
@@ -323,6 +323,7 @@ tr {
 td, th {
     display: table-cell;
     vertical-align: inherit;
+    padding: 1px;
```
To note that the correct fix is more along the lines of:

```
td:not(table td),
th:not(table th) {
    padding: 1px
}
```

karlcow replied:
hmm maybe not I read too fast. Let's check what the c++ part of the code does.

karlcow reviewed (commented):
The change is probably not the correct one now. It's why it is breaking a lot of tests.

karlcow commented:
with `run-webkit-tests --no-build imported/w3c/web-platform-tests/css/css-tables/`

```
Regressions: Unexpected text-only failures (10)
  imported/w3c/web-platform-tests/css/css-tables/colspan-001.html [ Failure ]
  imported/w3c/web-platform-tests/css/css-tables/colspan-002.html [ Failure ]
  imported/w3c/web-platform-tests/css/css-tables/colspan-003.html [ Failure ]
  imported/w3c/web-platform-tests/css/css-tables/height-distribution/computing-row-measure-0.html [ Failure ]
  imported/w3c/web-platform-tests/css/css-tables/height-distribution/computing-row-measure-1.html [ Failure ]
  imported/w3c/web-platform-tests/css/css-tables/html-to-css-mapping-1.html [ Failure ]
  imported/w3c/web-platform-tests/css/css-tables/percent-width-ignored-001.tentative.html [ Failure ]
  imported/w3c/web-platform-tests/css/css-tables/percent-width-ignored-002.tentative.html [ Failure ]
  imported/w3c/web-platform-tests/css/css-tables/percent-width-ignored-003.tentative.html [ Failure ]
  imported/w3c/web-platform-tests/css/css-tables/width-distribution/computing-column-measure-0.html [ Failure ]

Regressions: Unexpected image-only failures (1)
  imported/w3c/web-platform-tests/css/css-tables/colspan-004.html [ ImageOnlyFailure ]
```

It fully regresses more things than it solves. So there are other things to solve in the code than just this.

karlcow commented:
If we adopt the same hack than Chromium here. 

```
diff --git a/Source/WebCore/css/html.css b/Source/WebCore/css/html.css
index 1728b2a40e25..b1db4351db79 100644
--- a/Source/WebCore/css/html.css
+++ b/Source/WebCore/css/html.css
@@ -321,6 +321,11 @@ td, th {
     vertical-align: inherit;
 }
 
+td:not(table td),
+th:not(table th) {
+    padding: 1px
+}
+
 th {
     font-weight: bold;
     text-align: -internal-th-center;
```


There is one expected progression, and 0 regression.

```
Regressions: Unexpected text-only failures (1)
  imported/w3c/web-platform-tests/css/css-tables/html-to-css-mapping-1.html [ Failure ]
```

So there is something else in the code which enter into play here.

Ahmad-S792 commented:
I did try it in the past and had feedback from @nt1m , I will try to dig up PR and share here. It was bit more involved and we don't want to do what Chromium / Blink did and what to match specs.

karlcow commented:
yes I think we should find in the code what is creating this state with the normal css.

karlcow commented:
1. Go to about:blank
2. Insert the following in the Web Inspector console:
`document.body.appendChild( document.createElement('td') ) && document.body.firstElementChild;`
3. `window.getComputedStyle(document.querySelector('td')).padding`

* Expected: `1px`
* Actual: `0px`

Firefox and Chrome return `1px`
Safari returns `0px`

Ahmad-S792 commented:
My theory based on Blink's patch:

We have implementation in C++ to treat `parentTable` separately similar to Blink in `HTMLTableCellElement::additionalPresentationalHintStyle()`, which calls `additionalCellStyle()` for parent table. In `additionalCellStyle()`, we have call to `HTMLTableElement::additionalCellStyle()`, which creates sharedCell and adds following style `createSharedCellStyle()`, in which we have:

```
if (m_padding)
        style->setProperty(CSSPropertyPadding, CSSPrimitiveValue::create(m_padding, CSSUnitType::CSS_PX));
```

Where `m_padding` is `1` in `unsigned short m_padding { 1 };` in `HTMLTableElement.h`.

So we are dealing with `parent table` correctly and adding `1` padding so we have to modify style of UA stylesheet rules to exclude parent table (what Blink / Chromium did).

Although I agree with Karl's question as well about what does setting `0` to reflect.
